File: bridge/src/geolocation.ts
export interface Position {
  latitude: number
  longitude: number
  accuracy?: number
}

export interface PositionOptions {
  enableHighAccuracy?: boolean
  timeout?: number
  maximumAge?: number
}

function supportsBrowserGeolocation(): boolean {
  return typeof navigator !== 'undefined' && 'geolocation' in navigator
}

/**
 * Geolocation is a pure web feature on every platform — there is no Go-side
 * implementation to try first, so this calls navigator.geolocation directly.
 *
 * It used to `invoke('goleo:geolocationGetCurrentPosition')` and fall back to the
 * browser on failure. That handler no longer exists: of six platforms only Windows
 * ever had a native path, and it launched a PowerShell subprocess per call. The
 * webview reaches the same OS APIs itself, with the real permission UI.
 *
 * The app must still call `runtime.RegisterGeolocation` in Go. It registers no
 * handler, but it is what declares Android's ACCESS_FINE_LOCATION and iOS's
 * NSLocationWhenInUseUsageDescription — without which the WebView's own
 * geolocation request is denied.
 *
 * Rejects rather than inventing a position: a plausible-looking wrong coordinate
 * is worse than a failure the caller can handle.
 *
 * Only works while the page is alive and foregrounded. Background location is not
 * reachable through the web API.
 */
export function getCurrentPosition(options?: PositionOptions): Promise<Position> {
  return new Promise((resolve, reject) => {
    if (!supportsBrowserGeolocation()) {
      reject(new Error('geolocation not available'))
      return
    }
    navigator.geolocation.getCurrentPosition(
      (pos) => {
        resolve({
          latitude: pos.coords.latitude,
          longitude: pos.coords.longitude,
          accuracy: pos.coords.accuracy,
        })
      },
      (err) => reject(err),
      {
        enableHighAccuracy: options?.enableHighAccuracy ?? false,
        timeout: options?.timeout ?? 10000,
        maximumAge: options?.maximumAge ?? 0,
      },
    )
  })
}

File: runtime/geolocation/geolocation.go
//go:build !(android || ios) || goleo_geolocation

package geolocation

import "sync"

type Position struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
	Accuracy  float64 `json:"accuracy,omitempty"`
}

type PositionOptions struct {
	EnableHighAccuracy bool `json:"enableHighAccuracy,omitempty"`
	Timeout            int  `json:"timeout,omitempty"`    // ms
	MaximumAge         int  `json:"maximumAge,omitempty"` // ms
}

// Provider is a native geolocation backend. On mobile the shell registers
// one via SetProvider (FusedLocationProvider / CLLocationManager); on
// desktop the built-in platform implementation is used when no provider is
// set.
type Provider interface {
	GetCurrentPosition(opts PositionOptions) (*Position, error)
}

var (
	providerMu sync.RWMutex
	provider   Provider
)

func SetProvider(p Provider) {
	providerMu.Lock()
	defer providerMu.Unlock()
	provider = p
}

func GetCurrentPosition(opts PositionOptions) (*Position, error) {
	providerMu.RLock()
	p := provider
	providerMu.RUnlock()
	if p != nil {
		return p.GetCurrentPosition(opts)
	}
	return platformGetCurrentPosition(opts)
}
